cmd/mt-whisper-importer-writer/main.go

```diff
@@ -0,0 +1,289 @@
+	}
+	ttlTables := mdata.GetTTLTables(ttls, *windowFactor, mdata.Table_name_format)
+
+	p, err := partitioner.NewKafka(*partitionScheme)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to instantiate partitioner: %q", err))
+	}
+
+	server := &Server{
+		Cluster:     cassCluster,
+		Session:     session,
+		TTLTables:   ttlTables,
+		Partitioner: p,
+		Index:       cassandra.New(),
+	}
+	cluster.Init("mt-whisper-importer-writer", GitHash, time.Now(), "http", int(80))
+	server.Index.Init()
+
+	http.HandleFunc(*uriPath, server.chunksHandler)
+
+	log(fmt.Sprintf("Listening on %q", *httpEndpoint))
+	err = http.ListenAndServe(*httpEndpoint, nil)
+	if err != nil {
+		panic(fmt.Sprintf("Error creating listener: %q", err))
+	}
+}
+
+func throwError(msg string) {
+	msg = fmt.Sprintf("%s\n", msg)
+	if *exitOnError {
+		panic(msg)
+	} else {
+		printLock.Lock()
+		fmt.Fprintln(os.Stderr, msg)
+		printLock.Unlock()
+	}
+}
+
+func log(msg string) {
+	if *verbose {
+		printLock.Lock()
+		fmt.Println(msg)
+		printLock.Unlock()
+	}
+}
+
+func (s *Server) chunksHandler(w http.ResponseWriter, req *http.Request) {
+	metric := &archive.Metric{}
+	err := metric.UnmarshalCompressed(req.Body)
+	if err != nil {
+		throwError(fmt.Sprintf("Error decoding metric stream: %q", err))
+		return
+	}
+	log("Handling new metric")
+
+	if len(metric.Archives) == 0 {
+		throwError("Metric has no archives")
+		return
+	}
+
+	avg := false
+	if whisper.AggregationMethod(metric.AggregationMethod) == whisper.AggregationAverage {
+		avg = true
+	}
+
+	partition, err := s.Partitioner.Partition(&metric.MetricData, int32(*numPartitions))
+	err = s.Index.AddOrUpdate(&metric.MetricData, partition)
+	if err != nil {
+		throwError(fmt.Sprintf("Error updating metric index: %q", err))
+		return
+	}
+
+	for archiveIdx, a := range metric.Archives {
+		archiveTTL := a.SecondsPerPoint * a.Points
+		tableTTL, err := s.selectTableByTTL(archiveTTL)
+		if err != nil {
+			throwError(fmt.Sprintf("Failed to select table for ttl %d in %+v: %q", archiveTTL, s.TTLTables, err))
+			return
+		}
+		entry, ok := s.TTLTables[tableTTL]
+		if !ok {
+			throwError(fmt.Sprintf("Failed to get selected table %d in %+v", tableTTL, s.TTLTables))
+			return
+		}
+		tableName := entry.Table
+
+		if !avg || archiveIdx == 0 || !*fakeAvgAggregates {
+			log(fmt.Sprintf(
+				"inserting %d chunks of archive %d with ttl %d into table %s with ttl %d and key %s",
+				len(a.Chunks), archiveIdx, archiveTTL, tableName, tableTTL, a.RowKey,
+			))
+			s.insertChunks(tableName, a.RowKey, tableTTL, a.Chunks)
+		} else {
+			// averaged archives are a special case because mt doesn't store them as such.
+			// mt reconstructs the averages on the fly from the sum and cnt series, so we need
+			// to generate these two series out of raw averaged data by multiplying each point
+			// with the aggregation span and storing the result as sum, cnt is the aggregation span.
+			type sumCntChunks struct {
```

Comment:
this type is not used anywhere. (protip: get this tool https://github.com/dominikh/go-tools/tree/master/unused )